omigami/feature_selector.py

```diff
@@ -44,46 +53,144 @@ def __init__(
+        size, n_features = X.shape[0], X.shape[1]
+        groups = self.get_groups(groups, size)
+        input_data = InputData(X=X, y=y, groups=groups, n_features=n_features)
+        self.feature_evaluator.n_initial_features = n_features
+
+        repetition_results = Repetition()
+
+        for _ in range(self.repetitions):
+            data_splitter = DataSplitter(
+                self.n_outer,
+                self.n_inner,
+                input_data,
+                self.random_state,
+            )
+
+            for outer_split in data_splitter.iter_outer_splits():
+                outer_loop_results = self._run_outer_loop(
+                    input_data, data_splitter, outer_split
+                )
+                repetition_results.append(outer_loop_results)
+
+        # outer_loop = self._make_outer_loop(input_data)
+        # self._results = self._execute_repetitions(outer_loop)
+        self.selected_features = self.post_processor.select_features(repetition_results)
         self.is_fit = True
         return self
 
-    def _execute_repetitions(self, outer_loop: OuterLoop) -> List[Repetition]:
-        results = []
-        for _ in range(self.repetitions):
-            outer_loop.refresh_splits()
-            result = outer_loop.run(executor=self.executor)
-            results.append(result)
-        return results
+    def get_groups(self, groups: NumpyArray, size: int):
+        if groups is None:
+            logging.info("groups is not specified: i.i.d. samples assumed")
+            groups = np.arange(size)
+        return groups
+
+    def _run_outer_loop(
+        self, input_data: InputData, data_splitter: DataSplitter, outer_split: Split
+    ) -> OuterLoopResults:
+
+        raw_feature_elim_results = {}
+        feature_set = list(range(input_data.n_features))
+
+        while len(feature_set) >= self._minimum_features:
+            inner_results = []
 
-    def _make_feature_evaluator(self, input_data: InputData) -> FeatureEvaluator:
-        return FeatureEvaluator(
+            for inner_split in data_splitter.iter_inner_splits(outer_split):
+                inner_loop_data = input_data.split_data(inner_split, feature_set)
+                inner_results.append(
+                    self.feature_evaluator.evaluate_features(
+                        inner_loop_data, feature_set
+                    )
+                )
+
+            raw_feature_elim_results[tuple(feature_set)] = inner_results
+            feature_set = self._remove_features(feature_set, inner_results)
+
+        feature_elimination_results = self.post_processor.process_feature_elim_results(
+            raw_feature_elim_results
+        )
+        outer_loop_results = self.create_outer_loop_results(
+            feature_elimination_results, input_data, outer_split
+        )
+
+        return outer_loop_results
+
+    def _remove_features(self, features: List[int], results: InnerLoopResults):
```

Comment:
Is it supposed to stay here? Would have preferred to have this as class with an iterator so it is consistent with data_splitter but won't insist .

Reply:
We had it in a separate class initially but we thought it made the code clear having the while loop explicit. This meant the class only had the remove features method so we removed the class.